### PR TITLE
MAT-10060: update spring boot (from 3.5.14 to 4.0.7)

### PR DESCRIPTION
**Pull request by etan-sb**
## MADiE PR

Jira Ticket: [MAT-10060](https://jira.cms.gov/browse/MAT-10060)
(Optional) Related Tickets:

### Summary

# Spring Boot 4 + Jackson 3 Upgrade Notes — measure-service

Part of the fleet-wide SB 3.5.14 → 4.0.6 + Jackson 2 → 3 migration. Common patterns live in
`madie-fhir-service/SPRING_BOOT_4_UPGRADE.md`; this file records only what was specific to this repo.
measure-service is the **largest** repo (~33 Jackson files, 7 `@WebMvcTest` slices, 1747 tests) and surfaced
several brand-new fleet-wide issues — most importantly a java-models Jackson-3 regression (see below).

## pom.xml
- Parent `3.5.14` → `4.0.6`. okta `3.0.7` → `3.1.0`. java-models `0.9.6` → `0.10.0-SNAPSHOT`,
  rest-commons `1.0.6` → `1.1.0-SNAPSHOT`.
- `spring-boot-starter-web` → `spring-boot-starter-webmvc`; added `spring-boot-restclient`,
  `spring-boot-starter-security-oauth2-resource-server`/`-client`, test `spring-boot-webmvc-test`.
- **springdoc `2.8.10` → `3.0.3`.** springdoc 2.8.x is built for Spring Boot 3 and its
  `SpringDocHateoasConfiguration` references `org.springframework.boot.autoconfigure.hateoas.HateoasProperties`,
  which SB4 **relocated** to `org.springframework.boot.hateoas.autoconfigure.HateoasProperties` (new
  `spring-boot-hateoas` module) → every context that has both springdoc and `spring-boot-starter-hateoas`
  fails to load with `ClassNotFoundException: …hateoas.HateoasProperties`. springdoc **3.0.x** targets SB4.
  (measure-service is the only repo with springdoc + hateoas together.)
- **Removed the `junit-jupiter.version` 5.10.3 pin** (property + 4 `dependencyManagement` overrides that
  forced `junit-jupiter-*` to 5.10.3, originally added "to fix mongock versions"). Under SB4 the parent BOM
  brings **JUnit 6**; pinning Jupiter 5.10.3 against the JUnit-Platform-6 launcher throws
  `NoSuchMethodError: ExtensionContext$Store.computeIfAbsent(Object, Function, Class)` in every test. Let the
  BOM manage JUnit. (The plain `<dependencies>` junit-jupiter entries stay — they're now BOM-versioned.)

## ⚠️ java-models Jackson-3 regression — underscore-prefixed properties (fleet-wide, fixed in java-models)
- Lombok generates `get_type()`/`get_id()` for the `_type`/`_id` fields on the QDM data-criteria models.
  Jackson 3's stricter `DefaultAccessorNamingStrategy` only treats `get…`/`set…`/`is…` as accessors when the
  next char is an **uppercase letter**; `_` fails, so the getter AND setter are unrecognized, and with the
  fields `private`, the property **vanishes in both directions** (read drops it, write omits it). Jackson 2
  accepted those getters. This silently corrupts the QDM `_type` discriminator and the Mongo `_id`.
- **Fix lives in `madie-java-models`:** added `@JsonProperty("_type")` to the 56 `_type` fields and
  `@JsonProperty("_id")` to the 4 `_id` fields (explicit annotation pins the name and forces the field to be a
  property, bypassing the unrecognized accessors). Rebuilt java-models. This was invisible to every
  already-"done" repo because their tests never round-trip-and-assert `_type`/`_id`.
- **Same class of bug, local DTO:** `LockInfo` had `private boolean isLocked` (getter `isLocked()` →
  property `locked`; field `isLocked`). J3 **wrote `"locked"` but read expecting `"isLocked"`** → the boolean
  came back null → `Cannot map null into type boolean`. Fixed with `@JsonProperty("locked")` on the field
  (preserves the existing `locked` wire format and the Lombok `.isLocked(...)` builder/getter). Watch any
  `is`-prefixed boolean **field** name.

## Jackson 3 — source
- Mechanical: `com.fasterxml.jackson.{databind,core}` → `tools.jackson.{databind,core}`;
  `JsonProcessingException` → unchecked `JacksonException` (dead `catch (IOException)` around `readTree`/
  `readValue` → `catch (JacksonException)`, drop the now-unused `java.io.IOException` import). `@JsonSerialize`/
  `@JsonDeserialize` (on `MeasureListDTO.version`) → `tools.jackson.databind.annotation.*`; core annotations
  (`@JsonProperty`, `@JsonTypeInfo`, `@JsonSubTypes`, `@JsonInclude`) STAY on `com.fasterxml.jackson.annotation`.
- **`JsonNode.fields()` GONE** (`JsonUtil.replaceNestedDateTimeStringValue`) → `node.properties().forEach(...)`;
  guarded `asText()` to run only inside the `isTextual()` branch (J3 `asText()` is strict on container nodes).
  **`JsonNode.elements()` GONE** (`CqlDefinitionReturnTypeService`) → `node.values()`.
- `QdmTestCaseShiftDatesService`: `new ObjectMapper().registerModule(new JavaTimeModule())
  .setSerializationInclusion(NON_NULL)` (immutable mapper / `JavaTimeModule` removed) →
  `JsonMapper.builder().changeDefaultPropertyInclusion(incl -> incl.withValueInclusion(NON_NULL)).build()`
  (jsr310 auto-registered).
- `UserServiceClientConfig`: dropped the `MappingJackson2HttpMessageConverter` surgery → bare
  `new RestTemplateBuilder().additionalInterceptors(...).build()`; `RestTemplateBuilder` import →
  `org.springframework.boot.restclient`.
- `ErrorHandlingControllerAdvice`: `ErrorAttributes` → `org.springframework.boot.webmvc.error`
  (`ErrorAttributeOptions` stays on `org.springframework.boot.web.error`).
- `UriComponentsBuilder.fromHttpUrl(...)` → `fromUriString(...)` (`FhirServicesClient`, `ElmTranslatorClient`).
- `HttpClientErrorException.getRawStatusCode()` removed (`TestCaseValidationService`) → `getStatusCode().value()`.

## ⚠️ Jackson 3 `FAIL_ON_NULL_FOR_PRIMITIVES` now defaults to **true**
- `MeasureSearchCriteria` has `@AllArgsConstructor`/`@NoArgsConstructor` + two `boolean` primitives. With the
  parameter-names module (auto-registered in SB4), Jackson 3 picks the **all-args constructor** as the creator;
  a partial request body like `{}` then passes `null` for the absent primitives → 400
  `Cannot map null into type boolean`. Jackson 2 silently defaulted to `false`. The endpoints legitimately
  accept partial JSON, so restored the lenient behavior globally in `application.yml`:
  `spring.jackson.deserialization.fail-on-null-for-primitives: false`.

## MongoDB
- `spring.data.mongodb.{uri,database}` → `spring.mongodb.{uri,database}` across all 5 ymls; **`auto-index-creation`
  STAYS** on `spring.data.mongodb`. Test `@EnableAutoConfiguration(exclude=…)`:
  `MongoDataAutoConfiguration` → `DataMongoAutoConfiguration` (`…boot.data.mongodb.autoconfigure`);
  `MongoAutoConfiguration` → `…boot.mongodb.autoconfigure`.
- **⚠️ Hand-rolled `@Value` mirrors the yml key:** `MeasureMongoConfig` had
  `@Value("${spring.data.mongodb.uri}")` → must also become `${spring.mongodb.uri}` or the context fails
  `Could not resolve placeholder`. (Grep `@Value("${spring.data.mongodb` whenever you rename the yml keys.)

## Security
- Main `SecurityConfig` was already lambda-DSL → only added `@EnableWebSecurity`. **The test-only
  `SecurityConfigTest` (`@TestConfiguration` with a `filterChain(HttpSecurity…)` bean) ALSO needs
  `@EnableWebSecurity`** — without it the `@WebMvcTest` slices fail to load (`No qualifying bean HttpSecurity`).

## Tests
- `@WebMvcTest` package `…test.autoconfigure.web.servlet` → `org.springframework.boot.webmvc.test.autoconfigure`;
  `@MockBean`/`@SpyBean` → `@MockitoBean`/`@MockitoSpyBean`.
- **`@Captor` not initialized under SB4** (bean-override replaced the listener that used to call `openMocks`):
  added `@ExtendWith(MockitoExtension.class)` to the `@WebMvcTest` classes using `@Captor`
  (`MeasureControllerMvcTest`, `TestCaseControllerMvcTest`, `MeasureVersionControllerMvcTest`).
- **`@Mock` under `@ExtendWith(SpringExtension.class)` is null** (`ScoringPopulationValidatorTest`) →
  `@ExtendWith(MockitoExtension.class)` (pure unit test, no context needed).
- **Leaked-matcher cascade:** the null `@Mock` above made `when(repo.findGroupById(anyString()))` push a matcher
  then NPE before consuming it, leaving a dangling matcher that surfaced as `InvalidUseOfMatchers` in the *next*
  test class in the fork. Fixing the mock init cleared it — no change needed in the victim test.
- Mocking a Jackson parse failure: `new JsonParseException(mapper.getDeserializationContext().getParser(), …)`
  (J2 API, both gone) → `DatabindException.from((JsonParser) null, "…")`.
- `ResponseEntity.getStatusCodeValue()` removed → `getStatusCode().value()`.
- `new ObjectMapper().registerModule(new JavaTimeModule()).configure(WRITE_DATES_AS_TIMESTAMPS,false)` test
  helpers → `JsonMapper.builder().disable(DateTimeFeature.WRITE_DATES_AS_TIMESTAMPS).build()`
  (`SerializationFeature.WRITE_DATES_AS_TIMESTAMPS` relocated to `tools.jackson.databind.cfg.DateTimeFeature`).
- **Jackson 3 alphabetical key ordering** broke exact-JSON assertions. Fixed the **tests**, not production:
  JUnit `assertEquals(actual, exactJson)` → `JSONAssert.assertEquals(expected, actual, JSONCompareMode.STRICT)`
  (`MeasureControllerMvcTest`); MockMvc `content().string(exactJson)` → `content().json(exactJson,
  JsonCompareMode.STRICT)` (order-insensitive, `TestCaseControllerMvcTest`).
- `UserServiceClientConfigTest`: removed the two tests asserting the deleted `MappingJackson2HttpMessageConverter`
  surgery (obsolete).

## Result
`mvn clean install` green — 1747 tests, 0 failures.

## Runtime smoke-test findings (post-build)
- 🔴 **MongoDB driver-5.x UUID regression** (see playbook §6a): every `GET /api/measures/{id}` of a measure with a
  legacy `Binary` subtype-3 UUID (`testCases[].patientId`/`testCaseSetId`) 500s with
  `ConverterNotFoundException: Binary -> java.util.UUID` — the driver default moved to `UNSPECIFIED` and SD 5.x no
  longer sets it. **Fix:** added a `MongoClientSettingsBuilderCustomizer` bean in `MeasureMongoConfig` pinning
  `UuidRepresentation.JAVA_LEGACY`. Verified: 500→200 on real measures; 1747 tests green. (The `uuid-representation`
  property and a hand-built `@Bean` factory both fail — see §6a.)

### All Submissions
* [ ] This PR has the JIRA linked.
* [ ] Required tests are included.
* [ ] No extemporaneous files are included (i.e Complied files or testing results).
* [ ] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [ ] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/main/java/cms/gov/madie/measure/config/mongock/UpdateTestCaseJsonPatientUuidChangeUnit.java [hunk 1]
```diff
@@ -69,7 +69,7 @@ public void updatedTestCaseJsonWithPatientUuid(
 
   protected String updateJsonForUuids(
       final String measureId, TestCase testCase, TestCaseService testCaseService)
-      throws JsonProcessingException {
+      throws JacksonException {
```
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/config/mongock/UpdateTestCaseJsonPatientUuidChangeUnit.java]: this throws statement can be removed now that `JacksonException` is an unchecked exception.

File: src/main/java/cms/gov/madie/measure/utils/JsonUtil.java
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/utils/JsonUtil.java:1]: The `throws JacksonException` statement can be removed from all method signatures in this class.

The `catch` statements are still functional.

File: src/main/java/cms/gov/madie/measure/utils/TestCaseServiceUtil.java
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/utils/TestCaseServiceUtil.java:1]: The `throws JacksonException` statement can be removed from all method signatures in this class.

**Review by jkotanchik-SB — CHANGES_REQUESTED** [review]
There's actually quite a few files with methods throwing JacksonException. In all cases, we no longer need to explicitly state `throws JacksonException`, and the statement can be removed.

**Review by jkotanchik-SB — APPROVED** [review]
